### ISSUE TITLE
MPS backend produces corrupted output in SDXL image generation on macOS 27 beta; CPU output is correct (primitive ops match CPU)

### ISSUE THREAD

**Issue by talib79g-cmyk** [pytorch/pytorch#187280]
### 🐛 Describe the bug

After updating an Apple Silicon Mac to the macOS 27 ("Golden Gate") developer beta, PyTorch image generation (Stable Diffusion XL, via ComfyUI) on the MPS backend produces completely corrupted output — a uniform high‑frequency "noise / mush" image with no coherent structure. Running the identical workload on the CPU (--cpu) produces a correct, coherent image.

The regression appeared immediately on updating to the beta, with no change to PyTorch, the models, or the workflow. It worked correctly on macOS 26.

The corruption is independent of:

sampler / scheduler
prompt
attention backend (scaled_dot_product_attention path vs. split attention)
precision (default fp16/bf16 and forced fp32)

Only moving the computation to the CPU yields correct results.

What I was able to isolate

I compared isolated MPS ops against CPU reference outputs at half precision. scaled_dot_product_attention and conv2d, in both fp16 and bf16, match the CPU reference closely (no NaN/Inf):

fp16  attention relerr=0.0004  conv relerr=0.0004  nan/inf=False
bf16  attention relerr=0.0035  conv relerr=0.0029  nan/inf=False

(relerr = ‖mps − cpu‖ / ‖cpu‖, ref computed on CPU in fp32; tensors: attention q,k,v = randn(2,10,1024,64), conv x=randn(1,128,96,96), w=randn(128,128,3,3).)

So the fault does not appear to be in those primitives in isolation — it shows up only in the full model/sampler pipeline. I was not able to reduce it to a single-op minimal repro.

To Reproduce

Primary (what I observed):

Generate an SDXL text→image on MPS → corrupted noise.
Re-run the identical pipeline with the model on cpu → correct image.

Candidate minimal reproduction (⚠️ please run/confirm before relying on it — I could not fully verify which op triggers it):

pythonimport torch
from diffusers import StableDiffusionXLPipeline

model = "stabilityai/stable-diffusion-xl-base-1.0"
prompt = "a photograph of a dog in a park, detailed"

# MPS in fp16 (reproduces the corruption); CPU in fp32 as reference
for device, dtype in [("mps", torch.float16), ("cpu", torch.float32)]:
    pipe = StableDiffusionXLPipeline.from_pretrained(model, torch_dtype=dtype).to(device)
    img = pipe(prompt, num_inference_steps=20, height=512, width=512).images[0]
    img.save(f"out_{device}.png")
    print("saved", f"out_{device}.png")

Expected: out_mps.png ≈ out_cpu.png. Observed: out_mps.png is noise; out_cpu.png is correct.

Expected behavior

MPS output should match CPU output (coherent images), as on macOS 26.

Versions

Please paste the output of python -m torch.utils.collect_env here:

[run:  python -m torch.utils.collect_env  and paste the full output]

Known so far:

PyTorch: 2.10.0
Python: 3.11.14
OS: macOS 27.0 Developer Beta (build: ____ — from  → About This Mac)
Hardware: Mac mini (M4 Pro), 48 GB unified memory
torch.backends.mps.is_available() → True
App used to surface it: ComfyUI 0.21.1, SDXL checkpoint

### Versions

Additional context

Reproduces from a fresh process; not memory-pressure related (system memory pressure stayed green throughout).
Also filed with Apple via Feedback Assistant (Metal/MPS), since the root cause may be in the OS Metal framework rather than PyTorch.
Note: there is a separate, unrelated macOS 27 regression where host_statistics64(HOST_VM_INFO64) fails — that one is a macOS/psutil matter and has been reported to Apple; it is not a PyTorch bug and is mentioned only to avoid confusion if it appears in others' logs.

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

**Comment by jhavukainen** [comment]
Thanks for reporting @talib79g-cmyk . I'll try to reproduce it locally to see if I can track the source down. 

For future reference: You can use backtick symbols \` to do code formatting in the github issues to make it easier to understand. Surrounding a piece of text with a single pair of backticks gives you the inline code look like: `import torch` and surrounding a block of code or script with three pairs of backticks \`\`\` gives you a code block to isolate the repro script like:

```
import torch
from diffusers import StableDiffusionXLPipeline

model = "stabilityai/stable-diffusion-xl-base-1.0"
prompt = "a photograph of a dog in a park, detailed"

for device, dtype in [("mps", torch.float16), ("cpu", torch.float32)]:
  pipe = StableDiffusionXLPipeline.from_pretrained(model, torch_dtype=dtype).to(device)
  img = pipe(prompt, num_inference_steps=20, height=512, width=512).images[0]
  img.save(f"out_{device}.png")
  print("saved", f"out_{device}.png")
```

**Comment by talib79g-cmyk** [comment]
Thanks @jhavukainen, much appreciated — and noted on the backticks, thanks.

Here's my environment from `python -m torch.utils.collect_env`:

```
PyTorch version: 2.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 27.0 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.0.123.102)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.14 (main, Oct  9 2025, 16:16:55) [Clang 17.0.0 (clang-1700.6.3.2)] (64-bit runtime)
Python platform: macOS-27.0-arm64-arm-64bit
Is CUDA available: False
Is XPU available: False
Is XNNPACK available: True

CPU:
Apple M4 Pro

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] onnx==1.21.0
[pip3] onnxruntime==1.26.0
[pip3] pytorch-lightning==2.2.3
[pip3] torch==2.10.0
[pip3] torch-fidelity==0.3.0
[pip3] torchaudio==2.10.0
[pip3] torchmetrics==1.3.2
[pip3] torchsde==0.2.6
[pip3] torchvision==0.25.0
```

Hardware: Mac mini (M4 Pro), 48 GB unified memory. This is the macOS 27.0 (Golden Gate) developer beta; the same setup worked correctly on macOS 26.

(Note: ComfyUI runs in the venv shown under `[pip3]` — torch 2.10.0. collect_env also lists a separate conda base environment with torch 2.12.0 that I don't use for ComfyUI, in case that shows up.)

To summarise what I'm seeing: the corruption reproduces reliably in SDXL generation on the MPS backend, while the identical run on CPU is correct. It's independent of sampler, scheduler, prompt, attention backend, and precision (fp16/bf16 and forced fp32) — only CPU produces a correct image. Separately, isolated fp16/bf16 `scaled_dot_product_attention` and `conv2d` match the CPU reference closely (relerr < 0.004, no NaN/Inf), so it doesn't appear to be those primitives in isolation.

Happy to run your exact script and attach the resulting `out_mps.png` / `out_cpu.png` images if that would help confirm it on my setup — just let me know.

**Comment by talib79g-cmyk** [comment]
Here are the two outputs — same prompt, same fixed seed (12345), 512×512, ~20 steps, generated in ComfyUI. The only thing that differs between them is the compute device.

**CPU (correct):**

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/959fb76c-49a3-44f6-8579-6edac0c5a35b" />

**GPU / MPS (corrupted):**

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/0ffa07cc-f6b1-4258-83b1-7ac4c380f210" />

Same seed and composition in both runs — the dog and the park layout are still faintly visible in the MPS version, but the result is broken up into noise. So the GPU is attempting the same image and corrupting it, rather than producing something different.

**Comment by Isalia20** [comment]
is this only on macOS 27?

**Comment by talib79g-cmyk** [comment]
Yes, it was running perfectly on ios26.6 beta!!

**Comment by jhavukainen** [comment]
So I am able to verify the issue of MPS and CPU diverging on `torch==2.10` with macOS27 but it no longer reproduces with the newer torch. The working hypothesis at the moment is that 2.10 passed some op through MPSGraph that has now been moved to a metal kernel. The fixing PR seems to have landed at 2.12 (2.11 still produced diverging results).

@talib79g-cmyk are you able to update to a newer pytorch? The stable version 2.12 seems functional on my end.

**Comment by talib79g-cmyk** [comment]
Thanks @jhavukainen. I upgraded my ComfyUI environment to torch 2.12.0 (with torchvision 0.27.0, and torchaudio 2.11.0 since there's no 2.12 torchaudio yet), and confirmed from ComfyUI's startup log that it's actually running it:

```
pytorch version: 2.12.0
```

Unfortunately the corruption persists on my setup — a bare SDXL text-to-image workflow on MPS still comes out as noise/mush, while CPU is correct. So 2.12 hasn't resolved it here, even though it did on your end.

One thing worth flagging: I only ever verified this through ComfyUI, not through the minimal diffusers script — we never actually confirmed that script reproduces it on my machine. 

For reference: macOS 27.0 build Version 27.0 Beta (26A5353q), Mac mini (M4 Pro), 48 GB. I have now run your script.

Clear result, @jhavukainen: I ran your exact script on torch 2.12.0, and the MPS output is now correct — both MPS and CPU produce coherent images (attached). So your fix resolves the divergence at the PyTorch level on my machine too. Thank you for tracking it down.

The corruption I'm still seeing only occurs inside ComfyUI, so it must be hitting a different code path than plain diffusers/PyTorch — most likely its own compute backend or op routing rather than the op you fixed. So the remaining issue is ComfyUI-side, not PyTorch. I'll pursue it there.

Really appreciate the quick turnaround on this.

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/5bfbf45e-ab22-4283-b234-a819432c80b6" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/18832cee-176e-438f-b450-39a8ac9d0dd1" />

**Comment by jhavukainen** [comment]
Ok I had some time to take a deep dive to ComfyUI to figure out why wouldn't this work correctly and it turned out to be an adventure. I'll post the trajectory below but the summary is @malfet had some great foresight and fixed the problem already with the PR migrating argmin/argmax to Metal kernels https://github.com/pytorch/pytorch/pull/187304. 

Once that lands in the nightly @talib79g-cmyk (I checked that as of right now it was not in the nightly yet), please re-test running your workflow with that pytorch wheel to confirm so we can close the issue as solved. We've already addressed the underlying cause in macOS27 Beta MPSGraph as well so the issue will also go away with OS updates in due time. Here's my images before and after the fixing PR and below them is the explanation for the issue:

Before ComfyUI FP16 MPS:
<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/6f41cebb-c2ef-4666-b591-bcca5b4de16c" />

After PR #187304 ComfyUI FP16 MPS:
<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/030e42a6-f983-4334-871d-019455f248e5" />

### The issue
Turns out the issue is not fundamentally related to fp16 vs. fp32 implementation details but it's a side effect of a bug causing the argmin and argmax to round up the memory region it reads to nearest multiple of 1024. This causes the op to possibly return indices outside of the valid region if the bit pattern in that region mean a small (or large) value in the data type the op uses. This became an issue in the `timestep` function in sampling where ComfyUI adds noise to the diffused image according to a schedule:

comfy/model_sampling.py in ModelSamplingDiscrete.timestep() (L198–201):

```python     
  198:    def timestep(self, sigma):
  199:        log_sigma = sigma.log()
  200:        dists = log_sigma.to(self.log_sigmas.device) - self.log_sigmas[:, None]
  201:        return dists.abs().argmin(dim=0).view(sigma.shape).to(sigma.device)   # <-- broken MPS argmin
```

For fp32 in this program's runtime luck would have it so that the overflow region argmin looked into after the buffer containing the `dists` tensor data did not contain any bit patterns that would correspond to a smaller value than was in the valid data and the routine returns a valid index. For fp16 this was not the case and the smallest value ended up in the overflow region causing the timestep to get stuck at its maximum, so that a maximal amount of noise to be added at every step of the diffuser instead of gradually letting the noise die down to reach the converged output image. The reason the standalone script seemed to work fine has to do with this issue requiring the specific condition of having the overflow region contain a suitable bit pattern to trigger it, and probably slightly differing sampling implementation between the frameworks.

Thanks for filing the issue and the repro, this was very educational.

**Comment by talib79g-cmyk** [comment]
Confirmed fixed. Tested on torch nightly 2.14.0.dev20260624 (macOS 27 beta, M4 Pro, MPS), plain fp16, no workarounds: 3 clean SDXL/Pony "dog" runs plus 12 consecutive runs of a full Pony  workflow in one session — all coherent, no corruption on repeat runs. Thank you for the fast diagnosis and fix

**Comment by atalman** [comment]
This is cherry-picked. Hence closing this issue. Will be validating

**Comment by overlordxrz-source** [comment]
Can confirm this on macOS 27 beta with ComfyUI + SDXL.
**Environment**
- Hardware: MacBook Air M5, 16 GB unified memory
- OS: macOS 27.0 beta
- PyTorch: 2.9.0
- Python: 3.12
- App: ComfyUI 0.27.0
- Model: `sd_xl_base_1.0.safetensors`
- `torch.backends.mps.is_available()` → True
**Symptoms**
- MPS: SDXL txt2img at 1024×1024 produces structured noise / mush (no coherent image)
- CPU (`--cpu`): same workflow produces a correct image
- SD 1.5 on MPS works fine — issue appears SDXL-specific in practice, though the root cause seems backend-level
- Reproduces regardless of sampler (`dpmpp_2m`/`karras`, `euler`/`normal`), attention backend, and `--force-fp16` / `--force-fp32`
**What we isolated**
The failure traces to `argmin`/`argmax` on MPS returning out-of-bounds indices in ComfyUI's sigma/timestep lookup (`ModelSamplingDiscrete.timestep()` uses `.argmin(dim=0)`). On a tensor shaped like ComfyUI's sigma schedule, MPS can return an index equal to `shape[dim]` instead of a valid `0..shape[dim]-1` value. That silently feeds the denoiser wrong timestep embeddings every step, so sampling "completes" but never converges.
Minimal repro idea:
```python
import torch
x = torch.randn(1000, device="mps", dtype=torch.float32)
i_mps = x.argmin(dim=0)
i_cpu = x.cpu().argmin(dim=0)
print("mps", i_mps.item(), "cpu", i_cpu.item(), "valid range 0-999")
On our machine MPS sometimes returns 1000 where CPU returns a valid index in 0..999.

Workaround (until a fixed PyTorch build) Route argmin/argmax to CPU when the input tensor is on MPS, then move the result back. With that shim, SDXL on MPS produces correct images again.

Separate note (not PyTorch) We also hit intermittent host_statistics64(HOST_VM_INFO64) ... array not large enough from psutil on macOS 27 — unrelated to this image corruption, but it may show up in others' ComfyUI logs.

Happy to run additional tests if useful. Looks like #187304 may already address this — we're still on 2.9.0 so haven't verified the fixed build yet.

**Comment by talib79g-cmyk** [comment]
Confirmed fixed in stable 2.13.0: 6 consecutive SDXL/Pony runs (fp16, MPS, macOS 27 beta, M4 Pro, no workaround flags), all coherent and identical at fixed seed. Previously also confirmed on nightly 2.14.0.dev20260624. Thanks again — issue fully resolved on my end.
